keps/sig-node/2823-pod-addmission-plugin/README.md

```diff
@@ -0,0 +1,445 @@
+
+## Design Details
+
+### Design
+
+#### Configuration file
+
+Node-local Pod admission plugins will be listed in a configuration file. The
+Plugins field indicates the list of plugins to invoke before admitting the Pod.
+
+```
+{
+  "name": "admission-plugin",
+  "version": "0.1",
+  "plugins": [
+    {
+      "name": "sysctlcheck",
+      "type": "shell"
+    },
+    {
+      "name": "fargatecheck",
+      "type": "shell"
+    }
+  ]
+}
+```
+
+A node-local Pod admission plugin has the following structure:
+
+```
+
+// PluginType indicates type of the admission plugin
+type PluginType string
+
+const (
+    PluginTypeShell  PluginType = "shell"  // binary to execute.
+    PluginTypeGRPC   PluginType = "grpc"   // Local port on the host.
+    PluginTypeSocket PluginType = "socket" // fd to connect to.
+)
+
+// AdmissionPluginManager is the podAdmitHandler shim for external plugins.
+type AdmissionPluginManager struct {
+    confDir      string
+    binDir       string
+    pluginConfig *PluginConfig
+}
+
+// PluginConfig represents the plugin configuration file
+type PluginConfig struct {
+    Name         string              `json:"name,omitempty"`
+    Version      string              `json:"version,omitempty"`
+    Plugins      []*AdmissionPlugin  `json:"plugins"`
+}
+
+// AdmissionPlugin represents individual plugins specified in the configuration.
+type AdmissionPlugin struct {
+    Name     string     `json:"name"`
+    Type     PluginType `json:"type"`
+}
+
+```
+
+#### Feature gate
+
+This functionality adds a new feature gate named “PodAdmissionPlugin” which
+decides whether to invoke admission plugin or not.
+
+#### Kubelet to pod admission plugin communication
+
+Kubelet will encode the pod spec and invoke each admission plugin. After
+decoding the pod spec, plugins can perform additional validations and return the
+encoded form of the struct mentioned below to kubelet to decide whether to admit
+the pod or not. For shell plugin type, request will be sent over stdin and
+response will be received over stdout and stderr.
+
+```
+AdmitResult { 
+   Admit bool 
+   Reason string 
+   Message string 
+}
+```
+
+#### Implementation detail
+
+As part of this implementation, new sub package will be added to
+pkg/kubelet/lifecycle. In-tree admission handler shim will be included in this
+package, which will be responsible for discovering and invoking the pod
+admission plugins available on the host to decide on whether to admit the pod or
+not.
+
+If the plugin does not respond or if it's crashing, then Kubelet will fail-close
+and pod will not be accepted by the kubelet. Because if kubelet ignores the
+response and schedules the pod then intention of not executing pod with specific
+needs on these worker nodes will be violated. These plugins doesn't mutate the
+Pod object and can be invoked in parallel since there is no dependency between
+them.
+
+These validations will be applied to both static pods which are managed by
```

Comment:
It might be useful to have a mechanism to only run admission plugin on a subset of pods? Maybe we can have a way to exclude certain pods from such a check. WDYT?

Reply:
It is a nice idea. We would potentially like to visit it in the beta phase. If there is a use-case that you have, please let us know.